geopandas/sindex.py

```diff
@@ -1,25 +1,341 @@
+        """
+        Original rtree wrapper, kept for backwards compatibility.
+        """
+
+        def __init__(self, *args):
+            super().__init__(self, *args)
+
+        @property
+        def size(self):
+            return len(self.leaves()[0][1])
+
+        @property
+        def is_empty(self):
+            if len(self.leaves()) > 1:
+                return False
+            return self.size < 1
+
+    class RTreeIndex(rtree.index.Index):
+        """
+        A simple wrapper around rtree's RTree Index
+        """
+
+        def __init__(self, geometry):
+            stream = (
+                (i, item.bounds, idx)
+                for i, (idx, item) in enumerate(geometry.iteritems())
+                if pd.notnull(item) and not item.is_empty
+            )
+            try:
+                super().__init__(stream)
+            except RTreeError:
+                # What we really want here is an empty generator error, or
+                # for the bulk loader to log that the generator was empty
+                # and move on.
+                # See https://github.com/Toblerity/rtree/issues/20.
+                super().__init__()
+
+            # store reference to geometries for predicate queries
+            self._geometries = geometry.geometry.values
+            # create a prepared geometry cache
+            self._prepared_geometries = np.array([None] * self._geometries.size)
+
+        def query(self, geometry, predicate=None, sort=False):
+            """Compatibility layer for pygeos.query.
+
+            This is not an optimized function, if speed is important,
+            you are probably better off using PyGEOS.
+
+            Parameters
+            ----------
+            geometry : shapely geometry
+                A single shapely geometry to query against the spatial index.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                The predicate to use for testing geometries from the tree
+                that are within the input geometry's envelope.
+                See `pygeos.strtree` documentation for more details on
+                expected behavior of predicates.
+            sort : bool, default False
+                Determines if the results should be sorted. If False, results are
+                often sorted but there is no guarantee.
+
+            Returns
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+            if not geometry or geometry.is_empty:
+                return np.array([])
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_query = list(self.intersection(bounds, objects=False))
+
+            if not tree_query:
+                return np.array([])
+
+            # check predicate
+            if predicate in ("intersects", "within"):
```

Comment:
Like you say, the performance will depend on how many comparisons are being done. My thoughts were that we can cache the tree geometries but not the input geometries so this would be faster in the case where multiple input geometries match with the same tree geometries. But I guess it really depends on the datasets, I'm not sure that a general solution best for all datasets could be found.

Other things along these lines are (1) what geometry types to prep (as you mention) and (2) threshold preparing geometry depending on number of matches (currently, it will not prep for a single comparison).